File: backend/app/schemas/batch.py
"""
批量发送相关的 Pydantic Schemas
"""
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class BatchStatus(str, Enum):
    """批量任务状态"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class SmsBatchCreate(BaseModel):
    """创建批量任务"""
    batch_name: str = Field(..., min_length=1, max_length=200, description="批次名称")
    template_id: Optional[int] = Field(None, description="模板ID（使用模板时必填）")
    sender_id: Optional[str] = Field(None, max_length=20, description="发送方ID")
    send_config: Optional[Dict[str, Any]] = Field(default=None, description="发送配置")


class SmsBatchResponse(BaseModel):
    """批量任务响应"""
    id: int
    account_id: int
    batch_name: str
    template_id: Optional[int]
    file_path: Optional[str]
    file_size: Optional[int]
    total_count: int
    success_count: int
    failed_count: int
    processing_count: int
    # 按 SMSLog 实时聚合，与 SmsBatch.success_count（sent+delivered）口径拆分，便于与上游回执对照
    delivered_count: int = Field(0, description="状态为 delivered 的条数（终态回执：已送达）")
    sent_awaiting_receipt_count: int = Field(
        0,
        description="尚未终态送达的条数：sms_logs 中 pending+queued+sent（待发、队列中、已送通道待回执）",
    )
    status: BatchStatus
    error_message: Optional[str]
    sender_id: Optional[str]
    channel_code: Optional[str] = None
    message_preview: Optional[str] = Field(
        None,
        description="批次内首条 sms_logs.message 的样本，用于列表行内展示短信内容（前端截断显示）",
    )
    progress: int
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    @field_validator("batch_name", mode="before")
    @classmethod
    def batch_name_none_as_empty(cls, v: object) -> str:
        """极少数历史数据 batch_name 可能为 NULL"""
        if v is None:
            return ""
        return str(v)

    @field_validator(
        "total_count",
        "success_count",
        "failed_count",
        "processing_count",
        "progress",
        "delivered_count",
        "sent_awaiting_receipt_count",
        mode="before",
    )
    @classmethod
    def int_null_as_zero(cls, v: object) -> int:
        """旧数据 NULL 或 JDBC/部分驱动把整型以 str 返回时，避免响应校验 500"""
        if v is None:
            return 0
        if isinstance(v, bool):
            return int(v)
        if isinstance(v, int):
            return v
        if isinstance(v, float):
            return int(v)
        if isinstance(v, str):
            s = v.strip()
            if not s:
                return 0
            try:
                return int(float(s))
            except ValueError:
                return 0
        try:
            return int(v)  # Decimal 等
        except (TypeError, ValueError):
            return 0

    class Config:
        from_attributes = True


class SmsBatchListResponse(BaseModel):
    """批次列表响应"""
    total: int
    items: List[SmsBatchResponse]
    page: int
    page_size: int


class SmsBatchStats(BaseModel):
    """批次统计"""
    total_batches: int
    pending_batches: int
    processing_batches: int
    completed_batches: int
    failed_batches: int
    total_messages: int
    success_messages: int
    failed_messages: int


class BatchUploadResponse(BaseModel):
    """上传响应"""
    batch_id: int
    file_name: str
    file_size: int
    total_count: int
    message: str


class BatchRetryFailedResponse(BaseModel):
    """失败重发响应"""
    retried: int = Field(..., description="成功重新入队条数")
    skipped: int = Field(0, description="未重发条数（缺数据、计费或入队失败等）")
    pending_skipped: int = Field(0, description="因余额不足未重发的条数（已扣前 N 条仍然成功）")
    partial: bool = Field(False, description="是否部分成功（典型：余额仅够重发部分）")
    errors: List[str] = Field(default_factory=list, description="失败原因摘要（最多若干条）")
    message: str = Field(default="", description="说明")


class BatchResendUnsentResponse(BaseModel):
    """补发未发送响应"""
    triggered: bool = Field(..., description="是否已触发补发")
    unsent_count: int = Field(..., description="检测到的未发送条数")
    message: str = Field(default="", description="说明")

File: frontend/src/api/batch.ts
import { request } from './index'

export interface SmsBatch {
  id: number
  account_id: number
  batch_name: string
  template_id: number | null
  file_path: string | null
  file_size: number | null
  total_count: number
  success_count: number
  /** 回执终态：已送达（SMSLog delivered） */
  delivered_count?: number
  /** 通道已接受但仍为 sent：终态回执未到或解析失败（与上游门户可能暂时不一致） */
  sent_awaiting_receipt_count?: number
  failed_count: number
  processing_count: number
  status: 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled'
  error_message: string | null
  sender_id: string | null
  /** 批次内首条 sms_logs.message 样本（列表行内短信内容预览） */
  message_preview?: string | null
  progress: number
  started_at: string | null
  completed_at: string | null
  created_at: string
  updated_at: string
}

// 上传批量发送文件
export const uploadBatchFile = (data: FormData) => {
  return request.post('/batches/upload', data, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
}

// 获取批次列表
export const getBatches = (params?: { page?: number; page_size?: number; status?: string }) => {
  return request.get('/batches', { params })
}

// 获取批次统计
export const getBatchStats = () => {
  return request.get('/batches/stats')
}

// 获取批次详情
export const getBatchDetail = (id: number) => {
  return request.get(`/batches/${id}`)
}

// 取消批次
export const cancelBatch = (id: number) => {
  return request.delete(`/batches/${id}`)
}

/**
 * 失败记录重新计费并入队。
 *
 * partial=true 表示余额仅够重发前 N 条，已扣的不退；其余条目仍为 failed，
 * 行级 error_message 已写明「余额不足，未能重发」。前端应以 warning 提示。
 */
export const retryBatchFailed = (id: number, opts: { timeout?: number } = {}) => {
  // 新流程：生成新批次重发，正常计费正常路由，source_batch_id 溯源链
  return request.post<{
    success: boolean
    source_batch_id: number
    new_batch_id: number
    total_count: number
    async: boolean
    message: string
  }>(`/batches/${id}/retry-as-new-batch`, undefined, opts.timeout ? { timeout: opts.timeout } : undefined)
}

/** 补发未发送：处理 CSV 中尚未写入 sms_logs 的剩余号码（仅 CSV 上传批次可用） */
export const resendUnsentBatch = (id: number) => {
  return request.post<{ triggered: boolean; unsent_count: number; message: string }>(
    `/batches/${id}/resend-unsent`
  )
}

/**
 * 重新投递「排队中」记录到 Celery（不重复扣费）。
 * 适用于 Worker 异常导致 RabbitMQ 消息丢失、sms_log 卡 queued 的恢复。
 */
export const requeueBatchQueued = (id: number) => {
  return request.post<{
    retried: number
    skipped: number
    errors: string[]
    message: string
  }>(`/batches/${id}/requeue-queued`)
}

/** 导出批次明细 CSV（后端手机号脱敏） */
export const exportBatchRecordsCsv = (id: number) => {
  return request.get(`/batches/${id}/export`, { responseType: 'blob' })
}
